Components: Normalize displayed dates to UTC time for DateTimePicker

aduth opened the pull request:
## What?

Fixes #67480

Fixes an issue where differences between local browser and server timezones could cause incorrect dates to appear selected when choosing a date in the date picker.

Separately, fixes a minor TypeScript typings issue with the `offset` value in timezone settings, which are in-fact numeric ([source](https://github.com/WordPress/wordpress-develop/blob/6663b8d2f727525e819955f499e70d9fedc3fb6d/src/wp-includes/script-loader.php#L476)).

## Why?

Without these changes, there's a potential that posts could become published at a different time than expected.

## How?

Fixes the issue by normalizing displayed values to UTC. `DateTimePicker` values are without timezone (e.g. `2025-11-19T00:00:00`. Previously, these would be converted to browser local time before display, meaning that the options and numbers shown on the calendar may not actually align to the value saved for the post.

## Testing Instructions

Verify that picking a date around DST boundary selects the expected date:

1. Go to Posts > Add Post
2. Under "Post" settings in the sidebar, click "Publish: Immediately" option to select scheduled time
3. Change month to November 2025 (this should already be the case if you're reviewing this in November 2025)
4. Click "1" on the calendar view
5. Observe that the "Date" field shows "01 November 2025"

Repeat these steps in the following circumstance:

1. Change site "Timezone" to "New York" under Settings > General screen
2. Change your system timezone to GMT / London (in macOS this is [under "Date & Time" after de-selecting "Set time zone automatically"](https://github.com/user-attachments/assets/53e7b6c1-781d-4197-9d98-6d0f5d00d4cd))

Bonus points: Verify that the included regression tests fail on `trunk`, as a demonstration of the buggy behavior being effectively tested:

1. `git checkout trunk`
2. `git checkout fix/date-time-timezone-dst -- packages/components/src/date-time/date-time/test/index.tsx`
3. `npm run test:unit packages/components/src/date-time/date-time/test/index.tsx`
4. Observe that the tests fail

## Screenshots or screencast <!-- if applicable -->

The following screenshots show the result after clicking "1" in the scenario described above with GMT local time + EST server time.

Observe:

- Previously, clicking "1" would actually result in the saved value of "02" (see "Date" field)
- Previously, the calendar would show an option for October 31 despite not being in the current month.

|Before|After|
|-|-|
|<img width="345" height="506" alt="dst-before" src="https://github.com/user-attachments/assets/301735b3-e67f-46ea-ad56-21b2b7182663" />|<img width="349" height="511" alt="dst-after" src="https://github.com/user-attachments/assets/6b7e4356-ff20-4bb6-80f8-5abd477fba0a" />|

aduth committed:
Date: Fix incorrect typing on offset type

aduth committed:
DateTimePicker: Normalize displayed dates to UTC time

aduth commented on packages/components/src/date-time/date/index.tsx:
```diff
@@ -161,7 +155,7 @@ export function DatePicker( {
 			>
 				{ calendar[ 0 ][ 0 ].map( ( day ) => (
 					<DayOfWeek key={ day.toString() }>
-						{ dateI18n( 'D', day, -day.getTimezoneOffset() ) }
```
This can also be achieved by passing `true` as the third argument of `dateI18n`, but it's described as being "effectively deprecated". It's unclear but I don't get the impression that "effectively deprecated" extends to `gmtdateI18n` itself.

https://github.com/WordPress/gutenberg/blob/69251a1e17cc1efc430559960f5c037486d5e5fa/packages/date/src/index.ts#L493-L497

tyxla reviewed (commented):
This doesn't seem to work well with the default UTC+0 timezone:

https://github.com/user-attachments/assets/3f23e34a-d2ec-496c-a5fa-111dcd428e2b

See how the date in the field above differs from the selected date in the calendar.

aduth committed:
Parse timezoneless input string as UTC

aduth committed:
Add link to pull request in CHANGELOG.md

aduth commented on packages/components/src/date-time/utils.ts:
```diff
@@ -11,14 +12,17 @@ import type { InputAction } from '../input-control/reducer/actions';
 	if ( typeof input === 'string' ) {
-		return new Date( input );
+		// Check if the string ends with a timezone indicator per ISO-8601.
+		const hasTimezone = /Z|[+-]\d{2}(:?\d{2})?$/.test( input );
```
There's some similar patterns and usage in the date library, but (besides not being publicly exported) they're focused around checking just the timezone portion of the string and not an entire timezoneless string.

https://github.com/WordPress/gutenberg/blob/09eeecb837d3e8c5bfeb204892e7fc3a80130d3f/packages/date/src/index.ts#L22-L24

tyxla commented on packages/components/src/date-time/date-time/test/index.tsx:
```diff
@@ -0,0 +1,198 @@
+			screen.getByRole( 'button', {
+				name: 'November 15, 2025. Selected',
+			} )
+		).toBeInTheDocument();
```
Nit, especially for this test, but when testing visible items, I always prefer `toBeVisible()`, since `.toBeInTheDocument()` could be invisible to the user.

```suggestion
		).toBeVisible();
```

For negative queries, `.not.toBeInTheDocument()` is fine, since `isVisible()` will error.

aduth replied:
I had your voice in my head as I was including these, and came to the same conclusion about the negated assertions (we don't want them in the document, whether or not they're visible), but I guess I neglected to pay as much mind to the positive assertions 😅 I'll update.

aduth replied:
Updated positive assertions to use `toBeVisible` in ac743c6e9eb.

tyxla commented on packages/components/src/date-time/utils.ts:
```diff
@@ -11,14 +12,17 @@ import type { InputAction } from '../input-control/reducer/actions';
 	if ( typeof input === 'string' ) {
-		return new Date( input );
+		// Check if the string ends with a timezone indicator per ISO-8601.
+		const hasTimezone = /Z|[+-]\d{2}(:?\d{2})?$/.test( input );
```
Should we have unit tests for that util?

aduth replied:
> Should we have unit tests for that util?

Yeah, good idea 👍 It also led me down a rabbit hole of the types of date strings that JavaScript really supports (which is, in fact, _not_ all of ISO-8601). Updated in 972267f8639.

tyxla commented on packages/date/src/index.ts:
```diff
@@ -91,7 +91,7 @@ let settings: DateSettings = {
 		datetimeAbbreviated: 'M j, Y g: i a',
 	},
-	timezone: { offset: '0', offsetFormatted: '0', string: '', abbr: '' },
+	timezone: { offset: 0, offsetFormatted: '0', string: '', abbr: '' },
```
Can the change in type bite us in any way?

aduth replied:
This change is more accurate to the true behavior that happens in a real WordPress environment.

You can verify this on `trunk` that the number reported is numeric:

```
wp.date.getSettings().timezone.offset
// 0
```

These defaults are rarely used, as in WordPress they're overridden as part of the inline script for `wp-date` ([source](https://github.com/WordPress/wordpress-develop/blob/6663b8d2f727525e819955f499e70d9fedc3fb6d/src/wp-includes/script-loader.php#L476)).

As far as usage, it could have two effects, both of which could be perceived as a positive:

- We're usually converting this to a number anyways ([example](https://github.com/WordPress/gutenberg/blob/af72678d2e33a98c8fbe90b9854313112fbace6f/packages/components/src/date-time/time/timezone.tsx#L29), [example](https://github.com/WordPress/gutenberg/blob/af72678d2e33a98c8fbe90b9854313112fbace6f/packages/editor/src/components/post-schedule/label.js#L126)), where using a number instead of a string should have the same result. This is one of those cases where we probably don't even need to be doing this conversion if we could rely strictly on TypeScript types to enforce it as numeric.
- In cases where we're not, we're treating it as already numeric ([example](https://github.com/WordPress/gutenberg/blob/af72678d2e33a98c8fbe90b9854313112fbace6f/packages/editor/src/components/post-schedule/label.js#L119)), where this change could be _fixing_ some potential bugs

tyxla commented on package.json:
```diff
@@ -183,6 +183,7 @@
 		"stylelint-plugin-logical-css": "^1.2.3",
 		"terser": "5.32.0",
 		"terser-webpack-plugin": "5.3.10",
+		"timezone-mock": "1.3.6",
```
Why here and not in the components package where the test lives?

aduth replied:
I had a similar discussion in https://github.com/WordPress/gutenberg/pull/72847#discussion_r2479140529 where I highlighted that this has been the project convention up until recently with how we treat `devDependencies`. In that case, I think it made more sense for the `devDependencies` entry to exist at the package level. In this case, I feel like "mocking timezones" is something we could potentially want to support across any one of our packages' tests. I could also be convinced to take a hard stance that anything other than top-level scripts should have their dependencies defined in their own package, with the downside being that it may slightly disincentivize people from testing these kinds of edge cases.

aduth replied:
Shifted to the package's dependencies in 3b3d390da45 as the path of least resistance.

tyxla reviewed (commented):
This is testing well now, and I played with a bunch of DST scenarios and timezones. Nice. Just a few more questions and nitpicks 😉

aduth committed:
Add additional combination testing browser at UTC

matches original bug behavior, adds additional assurances in how we're using UTC as a baseline for formatting and normalizing dates

aduth committed:
Use toBeVisible over toBeInTheDocument with positive assertions

More accurate to the user's experience

aduth committed:
Add test coverage for inputToDate

aduth committed:
Define timezone-mock as devDependencies of components package

aduth commented:
> Do we still need a test without an offset (UTC+0), you know, just in case? It might be testing against the timezone mock library at first glance, but it would be ensuring we're catching a real bug here.

Yeah, I think it makes sense to include a case, since there could be some potential edge cases where the local browser time is already the same time as the timezone we're normalizing to (UTC). It's also pretty easy to add in how we're using `describe.each`. Updated in d17a3e4722b.

tyxla reviewed (approved):
Thanks @aduth, let's get it 🚢!
